Space key doesn't work when editing filters #1176

dariusf opened the pull request:
Fixes #1176.

dariusf commented:
@HansNewbie could you help to review? Thanks!

HansNewbie commented:
For the space testing, can we have some boundary tests as well? Maybe test adding space at start, end, and middle of a word that has no space or start and end with space? For example,
`count` -> `_count`
`count` -> `count_`
`count` -> `cou_nt`
`_count` -> `__count` -> `__count_` -> `__count__`

dariusf commented:
Added the comment.

Added the last one -- insertion of spaces when the field contains trailing spaces. That's significant because of the legacy code removed. The rest are all in the same class of boundary case and should be covered already, since spaces are not treated specially now.

HansNewbie commented on src/test/java/guitests/FilterTextFieldTest.java:
```diff
@@ -0,0 +1,136 @@
+        testSpaces(field);
+    }
+
+//    @Test
```
Is this meant to be commented? The `cancelTest` as well.

dariusf replied:
No, good thing you caught it. :sweat:

HansNewbie commented on src/main/java/ui/components/FilterTextField.java:
```diff
@@ -87,22 +88,39 @@ private void setup() {
+     * @return true if completion should be started
+     */
+    private boolean shouldStartCompletion() {
+        return getCharAfterCaret().trim().isEmpty() || !getSelectedText().isEmpty();
```
I need confirmation on this behaviour. What if the user select text in the middle of the filter? Should completion be performed? Say `assignee:me milestone:current` and the selected text is `ignee` of `assignee`.

dariusf replied:
With `ignee` selected, if you type `i`, the result is `assi[gnee]:me` (selection represented with square brackets). Typing anything else (say, `a`) results in `assa:me`. Completion is performed, but only if the current word + currently-typed letter is a prefix of some keyword.

The main reason for this change was to [not have completion cancelled as the user continues typing](https://github.com/HubTurbo/HubTurbo/blob/1167-space-key/src/test/java/guitests/FilterTextFieldTest.java#L22-L32). The other scenario of typing with some text already selected is a side effect.

HansNewbie replied:
Yep, I noticed about the completion toggling. Since this is an intended behaviour, sure. This is fine.

HansNewbie commented on src/main/java/filter/expression/QualifierType.java:
```diff
@@ -50,7 +50,7 @@
 
         defaultCompletions.addAll(Arrays.asList("closed", "open", "issue", "pr",
-            "pullrequest", "read", "unread",  "merged", "unmerged", "comments",
+            "pullrequest", "read", "unread", "me", "merged", "unmerged", "comments",
```
Does the alias `me` for `assignee` exist? I was using it in my previous comment for simplicity.

dariusf replied:
Ugh, just checked and it seems to have been removed at some point. It used to use your login name. Never mind then.

HansNewbie commented:
Ok seems good.
